File: common/Java/src/lib/lists/linkedlists/LinkedList.java
package lib.lists.linkedlists;

import lib.lists.List;
import nodes.ListNode;

public class LinkedList implements List {

    ListNode head;

    public LinkedList(ListNode head) {
        this.head = head;
    }

    @Override
    public final int size() {
        ListNode cur = this.head;
        int len = 0;
        while (cur != null) {
            len++;
            cur = cur.next;
        }
        return len;
    }

    public final int[] toArray() {
        ListNode cur = this.head;
        int[] res = new int[this.size()];

        int i = 0;
        while (cur != null) {
            res[i] = cur.val;
            cur = cur.next;
            i++;
        }

        return res;
    }

    public final ListNode getNodeByIndex(int idx) {
        ListNode cur = this.head;
        for (int i = 0; i < idx; i++) {
            cur = cur.next;
        }
        return cur;
    }
}

File: DSA/Cpp/src/lib/lists/linkedlist.cpp
#include "nodes.h"
#include "lib/lists/linkedlists.h"

namespace dsa
{
    namespace lib
    {
        LinkedList::LinkedList(nodes::ListNode *head) : head(head) {}
        void LinkedList::destroy()
        {
            nodes::ListNode *cur, *next;
            cur = this->head;
            while (cur)
            {
                next = cur->next;
                delete cur;
                cur = next;
            }
        };
        int LinkedList::size()
        {
            // ToDo
            return 0;
        };

        int *LinkedList::toArray()
        {
            // ToDo
            int *res;
            return res;
        }
        nodes::ListNode *LinkedList::getNodeByIndex(int idx) const {
            // ToDo
            nodes::ListNode *ls;
            return ls;
        }

        SinglyLinkedList::SinglyLinkedList(nodes::ListNode *head) : LinkedList(head){};

        // reverse the nodes before a certain node
        nodes::ListNode *SinglyLinkedList::reverseBefore(nodes::ListNode *node)
        {
            nodes::ListNode *pre, *cur, *next;
            pre = node;
            cur = this->head;

            while (cur != node)
            {
                next = cur->next;
                cur->next = pre;
                pre = cur;
                cur = next;
            }
            return pre;
        };

        nodes::ListNode *SinglyLinkedList::reverse()
        {
            nodes::ListNode *pre, *cur, *next;
            cur = this->head;

            while (cur)
            {
                next = cur->next;
                cur->next = pre;
                pre = cur;
                cur = next;
            }
            return pre;
        }

        // DoublyLinkedList::DoublyLinkedList(nodes::ListNode *head) : LinkedList(head){};

    } // namespace lib
} // namespace dsa